### PR TITLE
feat(search): Display warnings for on-demand metric queries

### PR DESCRIPTION
**Pull request by ArthurKnaus**
Display warnings for query filters that require on-demand metrics.

https://github.com/getsentry/sentry/assets/7033940/0c3a194c-63ea-42a3-9ef9-e5e42e4985d3

Pitfalls:
As the current HighlightQuery does not allow for tooltips to open on hover, there is no way to inspect the message if the query was rendered readonly. e.g. when opening a widget modal (see screenshot).

Screenshot 2023-08-14 at 16 57 15

Closes https://github.com/getsentry/team-ingest/issues/192

### CODE REVIEW

File: static/app/views/alerts/rules/metric/ruleConditionsForm.tsx [hunk 1]
```diff
@@ -386,6 +387,15 @@ class RuleConditionsForm extends PureComponent<Props, State> {
         []),
     ];
 
+    const getOnDemandFilterWarning = createOnDemandFilterWarning(
+      tct(
+        'We don’t routinely collect metrics from this property. However, we’ll do so [bold:once this alert has been saved.]',
+        {
+          bold: <Bold />,
```
Comment (matejminar) [inline static/app/views/alerts/rules/metric/ruleConditionsForm.tsx:394]:
```suggestion
          bold: <strong>,
```

**Review by matejminar — APPROVED** [review]
Looks good to me.

File: static/app/components/searchSyntax/renderer.tsx [hunk 2]
```diff
@@ -19,30 +19,49 @@ type Props = {
    * highlight active tokens and trigger error tooltips.
    */
   cursorPosition?: number;
+  /**
+   * A function that returns a warning message for a given filter key
+   * will only show a render a warning if the value is truthy
+   */
+  getFilterWarning?: (key: string) => React.ReactNode | undefined;
```
Comment (evanpurkhiser) [inline static/app/components/searchSyntax/renderer.tsx:26]: Would it make more sense to be provided the entire filter token?

Comment (ArthurKnaus) [reply static/app/components/searchSyntax/renderer.tsx:26]: agree, thx! 👍

**Review by evanpurkhiser — CHANGES_REQUESTED** [review]
Have some longer comments I'll leave on this

File: static/app/components/searchSyntax/renderer.tsx (hunk 2 above)
Comment (evanpurkhiser) [inline static/app/components/searchSyntax/renderer.tsx:26]: I'm not sure if this completely makes sense to be part of the `HighlightQuery`. I'm not a huge fan of this one-off-ish thing being added, especially since Filter tokens are already able to be marked `invalid`.

I am wondering if instead what we should do is add the `getFilterWarning` property to the `SearchConfig` object that is passed when parsing the search query.

https://github.com/getsentry/sentry/blob/6d355406cfe7e38eccb002f6f9aa9df2c66d19ec/static/app/components/searchSyntax/parser.tsx#L868-L871

I could imagine the function basically looks like this

```tsx
  /**
   * A function that returns a warning message for a given filter token
   * will only show a render a warning if the value is truthy
   */
  getFilterTokenWarning?: (keyName: string) => React.ReactNode;
```

(note, `React.ReactNode` already specifies undefined as one of it's types)

Then we add a new property to the resulting filter object type here:

https://github.com/getsentry/sentry/blob/6d355406cfe7e38eccb002f6f9aa9df2c66d19ec/static/app/components/searchSyntax/parser.tsx#L296-L323

Which could be like `warning: this.checkFilterWarning(filter, key, value)`

Which I imagine would just look like

```tsx
/**
 * Checks a filter against some non-grammar validation rules
 */
checkInvalidFilter = <T extends FilterType>(
  filter: T,
  key: FilterMap[T]['key'],
  value: FilterMap[T]['value']
) => {
  const keyName = getKeyName(key);

  return this.config.getFilterTokenWarning(keyName) ?? null;
}

Now `warning` has become a first class citizen on the filter type, so the `showWarning` variable inside of the `FilterToken` simply becomes

```
const showWarning = hasLeft && !!filter.warning;
```

Let me know what you think.

File: static/app/components/searchSyntax/parser.tsx [hunk 3]
```diff
@@ -666,6 +672,17 @@ export class TokenConverter {
...
 
+  /**
+   * Checks a filter against some non-grammar validation rules
+   */
+  checkFilterWarning = <T extends FilterType>(key: FilterMap[T]['key']) => {
+    if (key.type !== Token.KEY_SIMPLE) {
+      return null;
+    }
+    const keyName = getKeyName(key as TokenResult<Token.KEY_SIMPLE>);
+    return this.config.getFilterTokenWarning?.(keyName) ?? null;
```
Comment (ArthurKnaus) [inline static/app/components/searchSyntax/parser.tsx:683]: @evanpurkhiser I saw that `getKeyName` also supports `KEY_EXPLICIT_TAG` and `KEY_AGGREGATE`.
I am missing domain knowledge to fully understand those + afaik checking simple keys is enough for the current feature. Therefore, I would restrict this functionality to `Token.KEY_SIMPLE` for now. 🤔

File: static/app/components/searchSyntax/parser.tsx [hunk 4]
```diff
@@ -320,6 +320,11 @@ type FilterMap = {
      * The value of the filter
      */
     value: KVConverter<FilterTypeConfig[F]['validValues'][number]>;
+    /**
+     * A warning message associated with this filter
+     * The FE will render it as a tooltip
```
Comment (evanpurkhiser) [inline static/app/components/searchSyntax/parser.tsx:325]:
```suggestion
```

Let's leave out specifically what is done here, since that's a few abstraction levels up from what the searchSyntax parser's concerns

File: static/app/components/searchSyntax/parser.tsx (hunk 3 above)
Comment (evanpurkhiser) [reply static/app/components/searchSyntax/parser.tsx:683]: `KEY_EXPLICIT_TAG` looks like this: `tag[my_tag]:` which can also just be written as `my_tag` when it doesn't conflict with anything.

`KEY_AGGREGATE` is for things like `p99(transaction.duration):>50`

Comment (evanpurkhiser) [reply static/app/components/searchSyntax/parser.tsx:683]: Restricting to KEY_SIMPLE seems fime

File: static/app/components/searchSyntax/parser.tsx [hunk 5]
```diff
@@ -666,6 +672,17 @@ export class TokenConverter {
...
   };
 
+  /**
+   * Checks a filter against some non-grammar validation rules
+   */
+  checkFilterWarning = <T extends FilterType>(key: FilterMap[T]['key']) => {
+    if (key.type !== Token.KEY_SIMPLE) {
+      return null;
+    }
+    const keyName = getKeyName(key as TokenResult<Token.KEY_SIMPLE>);
```
Comment (evanpurkhiser) [inline static/app/components/searchSyntax/parser.tsx:682]: Doe the guard not correctly discriminate the key?

File: static/app/components/searchSyntax/renderer.spec.tsx
Comment (evanpurkhiser) [inline static/app/components/searchSyntax/renderer.spec.tsx:1]: Thanks for adding this test!

Could probably be it's own PR if you want to trim this one down to be specific to warnings for on-demand metric queries :-)

Either way is fine

File: static/app/components/smartSearchBar/index.tsx [hunk 6]
```diff
@@ -214,6 +214,11 @@ type Props = WithRouterProps &
      * as the stream view where it is a top level concept
      */
     excludedTags?: string[];
+    /**
+     * A function that returns a warning message for a given filter key
+     * will only show a render a warning if the value is truthy
+     */
+    getFilterWarning?: (key) => React.ReactNode | null;
```
Comment (evanpurkhiser) [inline static/app/components/smartSearchBar/index.tsx:221]:
```suggestion
    getFilterWarning?: (key) => React.ReactNode;
```

`ReactNode` already includes `null`

**Review by evanpurkhiser — APPROVED** [review]
Left a few comments, but I think it's good to go!

**Comment by evanpurkhiser** [comment]
Thanks for taking the time to clean it up :-)

File: static/app/components/searchSyntax/parser.tsx (hunk 5 above)
Comment (evanpurkhiser) [reply static/app/components/searchSyntax/parser.tsx:682]: ah yea, seems like the type is a little bit too complicated for TS to understand

File: static/app/utils/onDemandMetrics/index.tsx [hunk 7]
```diff
@@ -58,3 +72,14 @@ export const extrapolatedAreaStyle = {
   },
   opacity: 1.0,
 };
+
+export function hasOnDemandMetricAlertFeature(organization) {
```
Comment (obostjancic) [inline static/app/utils/onDemandMetrics/index.tsx:76]: nit: maybe just add type for the organization here

**Review by obostjancic — APPROVED** [review]
🚢

File: static/app/utils/onDemandMetrics/index.tsx (hunk 7 above)
Comment (ArthurKnaus) [reply static/app/utils/onDemandMetrics/index.tsx:76]: oh, right. Was not aware we even allow implicit any 😮
